Boat support, multiple passengers, "last dropped item" expression, new event values and fixes.

Tuke-Nuke opened the pull request:
**Changes**:
* Added boat support for `1.10+`:
  * Patterns: `boat, any boat, oak boat, spruce boat, birch boat, jungle boat, acacia boat, dark oak boat`
  * `boat` and `any boat` works like a supertype when comparing, but `boat` spawns a `oak boat` and `any boat` spawns a random boat in spawn effects.
  * Also added boat to the default comparator of entity's materials to work when compared "with items".
  * Boats will return it's type when returned as string, like `oak boat`.
* Added multiple passenger support for `1.11.2+`.
  * The syntax still the same, the difference is for 1.11.2+ will return a list instead of a single entity and can accept all changers (add, remove, set...).
  * For 1.11 and lower will be the same as previously: single entity and it can only be setted.
  * Changed the syntax of `EffVehicle` to accept multiple entities when supported.
  * Created a `PassengerUtils` to handle between versions, using reflection as the methods will be removed soon.
* Added a new syntax for `last spawned entity`: `[the] [last[ly]] dropped item`
  * It was necessary just to be more friendly syntax, it was a little weird with `last spawned dropped item`.
  * Also, added the experience that are dropped from `drop effect` with `last spawned experience orb` too. (Before It was only stored when using `spawn effect`).
* Fixed a bug when checking `if a entity is a armor stand` adding a default comparator for it.
* Added `event-item` to `item spawn` event. (I was pretty sure that it was already there since 2.1)

<details><summary>Examples and test codes</summary><p>

```
spawn a boat or oak boat #Both the same
spawn a spruce boat, birch boat, jungle boat, acacia boat, dark oak boat at player
spawn any boat #Randomly

if target is any boat or a boat: #Both works return true for all boats
    send "It's a boat"
if target is a oak boat:
    send "Oak boat: true"
else if target is a birch boat:
    send "Birch boat: true"
else if target is a spruce boat:
    send "Spruce boat: true"
else if target is a jungle boat:
    send "Jungle boat: true"
else if type of target is a acacia boat:
    send "Acacia boat: true"
else if target is a dark oak boat:
    send "Dark oak boat: true"
else:
    send "None of them"

spawn any boat
add a pig, a chicken, a creeper and a creeper to passengers of last spawned boat #It doesn't have limits.
wait a second
remove all creepers from passengers of last spawned boat
wait a second
clear passengers of last spawned boat

spawn any boat
make {_player1} and {_player2} ride last spawned boat

drop a stone without velocity
push last dropped item upwards

spawn a armor stand
if last spawned entity is a armor stand:
    send "true"
else:
   send "false"

on item spawn:
    broadcast "%event-item%"
```
</p></details>
<p></p>

I considered `dev26` as next version id, they are referenced at: [here](https://github.com/bensku/Skript/pull/438/files#diff-e78eb62873567060b0cde657767e8499R59) and [here](https://github.com/bensku/Skript/pull/438/files#diff-b7b981856cf73767fc903119c84cf5fcR67)

TheBentoBox commented:
So good! Now we just need someone to fix villagers being `entities.normals.name` and the block face feature request from #67 . ( @LimeGlass ?)

Tuke-Nuke commented:
@TheBentoBox I know how to fix the villager's name, also just a question. It's better to make it return the villager profession or just `villager` when converting a entity to string?

TheBentoBox commented:
For the sake of consistency and backwards compatibility (since it used to return just `villager` before it broke), it should return just `villager`. How it currently works in Skript is each entity returns its core entity type instead of its subtype, e.g. all normal horses are just `horse` instead of `brown horse with snowflake appaloosa marking` or something.

That being said, adding a profession type would be nice and would make sense, so we could use the `getProfession` and `getVillagerProfession` methods of Villager and ZombieVillager respectively. Something like a `[villager] profession of %living entity%` expression (which silently ignores other entity types) and accepts both setting and getting (probably not resetting or clearing as there isn't a clear "default state" villager, unless you want to make it default to the Nitwit profession as that's sort of the "generic" villager).

Adding the extra expression/type is up to you though, just making it return `villager` again instead of `entities.normals.name` would be great.

Tuke-Nuke commented:
I agree, so I will commit in some minutes to fix that, it's just a missing section called `normal` in `english.lang` file. About the expression, I will let for the next PR cause it needs a little time to test to make sure not conflict with anything.

TheLimeGlass commented:
Great pull request, the boats I'm not sure about because of lower versions.

Tuke-Nuke commented:
Thank you. It will work as well because It will only register the new boats if it does support it (for 1.10+), if it doesn't, it will use the old Skript method for boats. Same thing when [comparing entity](https://github.com/bensku/Skript/pull/438/files#diff-af85df5511dc42546a0bf46b7bdd1b3aR252), in case it doesn't support new boats, it will use previous method to do it.

TheBentoBox commented:
Sounds good, I agree it makes the most sense for the expression to be in a separate PR.

Tuke-Nuke commented:
Well, I added the two fixes that you suggestted in this PR already, since they were simple to do.
* Added `event-direction` for 2 events:
  * `on click`: It returns the side of `clicked block`, if it is avaliable.
  * `on place`: It returns the direction from `event-block` to `block that player placed on`. I added to this one too since it was possible too.
      * `block event-direction` can return `event-block` when the replaced block doesn't have a side and can be replaced when placing blocks, like placing a block in a `long grass`.
* Fixed villagers returning `entities.normal.name` when converting them to string.

**Examples**:
```
on place:
    set block event-direction to stone #It will return the block that you placed against to.

on click:
    if clicked block is set:
        if event-direction is north or south:
            send "idk"
```

TheBentoBox commented:
Really nice job, and quick work too! Especially love that implementation within the `on place` event, that's gonna be super useful for simulating placement events and stuff.

TheBentoBox commented:
@bensku I see you working on stuff still and labeling it as dev26, I'd make sure this gets merged in before then as it's all labeled with the plan of going into dev26 (and is all really good).

bensku commented:
Yep, will merge this before dev26. However, dev26 ETA is "not this week" since I need some time to finish new chat formatting and test it.

bensku commented:
Merging now. I hope that nothing breaks :)

Tuke-Nuke commented:
Adding entities is always complicated. One of things that I took care to not conflict with was items, since they have same name, but Skript itself didn't conflicted, some stuffs like this worked without problems
```
set {_item} to spruce boat #It gets the item
give spruce boat to player
give {_item} to player #Both ok
#Conditions
if {_item} is boat: #False, checked as item but not the same type. If change "boat" to refer to "any boat" will fix it
if {_item} is spruce boat: #True
if {_item} is any boat: #False, cause in Aliases it only refers to oak boat
if target is {_item}: #True if the entity is a spruce boat, as Skript has a comparator Entity == Item
```
